File: chap12/test-code/Test4/udp_chat_client.py
# udp_chat_client.py
import socket
import threading

SERVER_HOST = "localhost"   # Mac에서 실행하면 localhost -> 도커 서버로 포워딩
SERVER_PORT = 2600
BUF = 1024

def recv_loop(sock):
    while True:
        data, _ = sock.recvfrom(BUF)
        print("\n[MSG]", data.decode("utf-8", errors="replace"))
        print("> ", end="", flush=True)

def main():
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
        # 서버에게 첫 인사 보내서 자신의 주소 등록
        nickname = input("닉네임 입력: ")
        s.sendto(f"{nickname} 님 입장".encode("utf-8"), (SERVER_HOST, SERVER_PORT))

        # 수신 스레드 시작
        t = threading.Thread(target=recv_loop, args=(s,), daemon=True)
        t.start()

        while True:
            msg = input("> ")
            if msg.lower() == "exit":
                s.sendto(f"{nickname} 님 퇴장".encode("utf-8"), (SERVER_HOST, SERVER_PORT))
                break
            s.sendto(f"{nickname}: {msg}".encode("utf-8"), (SERVER_HOST, SERVER_PORT))

if __name__ == "__main__":
    main()

File: chap12/test-code/Test4/UdpChatClient.java
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.Scanner;

public class UdpChatClient {
    private String host;
    private int port;
    private int bufSize;
    private String nickname;

    public UdpChatClient(String host, int port, int bufSize, String nickname) {
        this.host = host;
        this.port = port;
        this.bufSize = bufSize;
        this.nickname = nickname;
    }

    public void start() throws Exception {
        DatagramSocket socket = new DatagramSocket();
        InetAddress serverAddr = InetAddress.getByName(host);

        // send join message
        String joinMsg = nickname + " joined the chat";
        byte[] join = joinMsg.getBytes("UTF-8");
        socket.send(new DatagramPacket(join, join.length, serverAddr, port));

        // receive thread
        Thread recvThread = new Thread(() -> {
            byte[] buf = new byte[bufSize];
            try {
                while (true) {
                    DatagramPacket p = new DatagramPacket(buf, buf.length);
                    socket.receive(p);
                    String msg = new String(p.getData(), 0, p.getLength(), "UTF-8");
                    System.out.println("\n[MSG] " + msg);
                    System.out.print("> ");
                }
            } catch (Exception ignored) {}
        });

        recvThread.setDaemon(true);
        recvThread.start();

        // send loop
        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.print("> ");
            String line = sc.nextLine();

            if (line.equalsIgnoreCase("exit")) {
                String byeMsg = nickname + " left the chat";
                byte[] bye = byeMsg.getBytes("UTF-8");
                socket.send(new DatagramPacket(bye, bye.length, serverAddr, port));
                break;
            }

            String fullMsg = nickname + ": " + line;
            byte[] out = fullMsg.getBytes("UTF-8");
            socket.send(new DatagramPacket(out, out.length, serverAddr, port));
        }

        socket.close();
        sc.close();
    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        System.out.print("Nickname: ");
        String nick = sc.nextLine();

        UdpChatClient client = new UdpChatClient("localhost", 2600, 1024, nick);
        client.start();
    }
}
